File: DesignPatterns/Behavioral/Strategy/Java/ShoppingCartClient.java
import java.util.ArrayList;
import java.util.List;

class Item {
  private String upcCode;
  private int price;

  public Item(String upcCode, int price) {
    this.upcCode = upcCode;
    this.price = price;
  }

  public String getUpcCode() {
    return this.upcCode;
  }

  public int getPrice() {
    return this.price;
  }
}

class ShoppingCart {
  private List<Item> items;

  public ShoppingCart() {
    this.items = new ArrayList<Item>();
  }

  public void addItem(Item item) {
    this.items.add(item);
  }

  public void removeItem(Item item) {
    this.items.remove(item);
  }

  public int calculateTotal() {
    int sum = 0;
    for(Item item: items){
      sum = sum + item.getPrice();
    }
    return sum;
  }

  public void pay(PaymentStrategy paymentMethod) {
    int amount = calculateTotal();
    paymentMethod.pay(amount);
  }
}

interface PaymentStrategy {
  public void pay(int amount);
}

class CreditCardStrategy implements PaymentStrategy {

  private String name;
  private String cardNumber;
  private String cvv;
  private String dateOfExpiry;

  public CreditCardStrategy(String name, String cardNumber, String cvv, String dateOfExpiry) {
    this.name = name;
    this.cardNumber = cardNumber;
    this.cvv = cvv;
    this.dateOfExpiry = dateOfExpiry;
  }

  public void pay(int amount) {
    System.out.println(amount + " paid with credit/debit card");
  }
}

class PaytmStrategy implements PaymentStrategy {

  private String email;
  private String password;

  public PaytmStrategy(String email, String password) {
    this.email = email;
    this.password = password;
  }

  public void pay(int amount) {
    System.out.println(amount + " paid via paytm");
  }
}

public class ShoppingCartClient {
  public static void main(String[] args) {
    ShoppingCart cart = new ShoppingCart();
    cart.addItem(new Item("1234", 10));
    cart.addItem(new Item("5678", 40));

    cart.pay(new PaytmStrategy("ravi@gmail.com", "pass1234"));
    cart.pay(new CreditCardStrategy("Ravi", "123456781234","786", "12/15"));
  }
}

File: DesignPatterns/Behavioral/Strategy/Ruby/shopping_cart_client.rb
class Item
  attr_reader :upc_code, :price

  def initialize(upc_code, price)
    @upc_code = upc_code
    @price = price
  end
end

class ShoppingCart
  def initialize
    @cart_items = []
  end

  def add_item(item)
    @cart_items << item
  end

  def remove_item(item)
    @cart_items.delete_if { |it| it == item }
  end

  def calculate_total
    total = 0
    @cart_items.each do |item|
      total += item.price
    end
    total
  end

  def pay(stratergy)
    amount = calculate_total
    stratergy.pay(amount)
  end
end

class Stratergy
  def pay(amount)
    raise NotImplementedError
  end
end

class CreditCardStratergy < Stratergy
  attr_reader :name, :card_number, :cvv, :date_of_expiry

  def initialize(name, card_number, cvv, date_of_expiry)
    @name = name
    @card_number = card_number
    @cvv = cvv
    @date_of_expiry = date_of_expiry
  end

  def pay(amount)
    puts "#{amount} paid with credit/debit card of name - #{@name}"
  end
end

class PaytmStratergy < Stratergy
  attr_reader :email, :password

  def initialize(email, password)
    @email = email
    @password = password
  end

  def pay(amount)
    puts "#{amount} paid with paytm of email - #{@email}"
  end
end

shopping_cart = ShoppingCart.new
item1 = Item.new("1234", 10)
item2 = Item.new("5678", 40)

shopping_cart.add_item(item1)
shopping_cart.add_item(item2)

shopping_cart.pay(PaytmStratergy.new("ravi@gmail.com", "pass1234"))
shopping_cart.pay(CreditCardStratergy.new("Ravi", "123456781234","786", "12/15"))
